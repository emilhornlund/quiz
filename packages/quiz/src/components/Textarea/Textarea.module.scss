@import '../../styles/colors';
@import '../../styles/helpers';

$outline-offset: pxToRem(2px);

.container {
  display: flex;
  flex-direction: column;
  margin: $outline-offset;
  height: 100%;
  width: 100%;
  max-width: 100%;
  flex: 1;

  textarea {
    display: flex;
    flex-direction: column;
    appearance: none;
    outline: $white solid $outline-offset;
    outline-offset: $outline-offset;
    border: none;
    width: 100%;
    height: 100%;
    resize: none;
    background-color: $white;
    color: $gray-4;

    @include mobile {
      padding: $layout-gutter * 0.8;
      border-radius: $layout-gutter * 0.5;

      @include fontSize(16px);
    }

    @include tablet {
      padding: $layout-gutter * 0.9;
      border-radius: $layout-gutter * 0.6;

      @include fontSize(17px);
    }

    @include desktop {
      padding: $layout-gutter;
      border-radius: $layout-gutter * 0.7;

      @include fontSize(18px);
    }
  }
}