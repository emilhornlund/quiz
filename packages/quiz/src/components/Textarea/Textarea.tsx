import React, { FC } from 'react'

import styles from './Textarea.module.scss'

export interface TextareaProps {
  id: string
  name?: string
  value?: string
  onChange?: (value: string) => void
}

const Textarea: FC<TextareaProps> = ({ id, name = id, value, onChange }) => {
  return (
    <div className={styles.container}>
      <textarea
        id={id}
        name={name}
        value={value}
        onChange={(event) => onChange?.(event.target.value)}
        data-testid={`test-${id}-textarea`}
      />
    </div>
  )
}

export default Textarea
