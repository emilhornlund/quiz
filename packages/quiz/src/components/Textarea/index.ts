export type { TextareaProps } from './Textarea'
export { default } from './Textarea'
