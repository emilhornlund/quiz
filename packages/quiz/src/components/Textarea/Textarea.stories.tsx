import type { Meta, StoryObj } from '@storybook/react'

import Textarea from './Textarea'

const meta = {
  component: Textarea,
  tags: ['autodocs'],
} satisfies Meta<typeof Textarea>

export default meta
type Story = StoryObj<typeof meta>

export const Default = {
  args: {
    id: 'my-text-field',
  },
} satisfies Story
