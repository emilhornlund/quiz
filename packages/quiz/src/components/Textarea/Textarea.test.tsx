import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { describe, expect, it, vi } from 'vitest'

import Textarea from './Textarea'

describe('Textarea', () => {
  it('should render a Textarea with default props', async () => {
    const { container } = render(<Textarea id="my-textarea" />)

    expect(container).toMatchSnapshot()
  })

  it('should fire onChange when Textarea receives text', async () => {
    const onChange = vi.fn()

    const { container } = render(
      <Textarea id="my-textarea" onChange={onChange} />,
    )

    const textarea = screen.getByTestId(
      'test-my-textarea-textarea',
    ) as HTMLTextAreaElement
    fireEvent.focus(textarea)
    fireEvent.change(textarea, { target: { value: 'some value' } })

    expect(textarea.value).toBe('some value')

    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith('some value')

    expect(container).toMatchSnapshot()
  })
})
