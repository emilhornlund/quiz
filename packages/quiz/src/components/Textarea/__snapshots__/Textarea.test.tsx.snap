// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Textarea > should fire onChange when Textarea receives text 1`] = `
<div>
  <div
    class="container"
  >
    <textarea
      data-testid="test-my-textarea-textarea"
      id="my-textarea"
      name="my-textarea"
    />
  </div>
</div>
`;

exports[`Textarea > should render a Textarea with default props 1`] = `
<div>
  <div
    class="container"
  >
    <textarea
      data-testid="test-my-textarea-textarea"
      id="my-textarea"
      name="my-textarea"
    />
  </div>
</div>
`;
